swap/swap.go

```diff
@@ -17,69 +17,340 @@
+	peerBalance := s.balances[peer]
+	err := s.stateStore.Put(peer.String(), &peerBalance)
+	if err != nil {
+		log.Error("error while storing balance for peer", "peer", peer.String())
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(peerBalance, 10))
+	return peerBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
+// logBalance is a helper function to log the current balance of a peer
+func (s *Swap) logBalance(peer *protocols.Peer) {
+	err := s.loadState(peer)
+	if err != nil && err != state.ErrNotFound {
+		log.Error(fmt.Sprintf("error while loading balance for peer %s", peer.String()))
+	} else {
+		log.Info(fmt.Sprintf("balance for peer %s is %d", peer.ID(), s.balances[peer.ID()]))
+	}
+}
+
+// sendCheque sends a cheque to peer
+func (s *Swap) sendCheque(peer enode.ID) error {
+	swapPeer := s.getPeer(peer)
+	cheque, err := s.createCheque(peer)
+	if err != nil {
+		log.Error("error while creating cheque: %s", err.Error())
+		return err
+	}
+
+	log.Info(fmt.Sprintf("sending cheque with serial %d, amount %d, benficiary %v, contract %v", cheque.ChequeParams.Serial, cheque.ChequeParams.Amount, cheque.Beneficiary, cheque.Contract))
+	s.cheques[peer] = cheque
+
+	err = s.stateStore.Put(peer.String()+"_cheques", &cheque)
+	// TODO: error handling might be quite more complex
+	if err != nil {
+		log.Error("error while storing the last cheque: %s", err.Error())
+		return err
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// reset balance;
+	// TODO: if sending fails it should actually be roll backed...
+	s.resetBalance(peer, int64(cheque.Amount))
+
+	err = swapPeer.Send(context.TODO(), emit)
 	return err
 }
 
-//GetPeerBalance returns the balance for a given peer
-func (swap *Swap) GetPeerBalance(peer enode.ID) (int64, error) {
-	swap.lock.RLock()
-	defer swap.lock.RUnlock()
-	if p, ok := swap.balances[peer]; ok {
-		return p, nil
+// Create a Cheque structure emitted to a specific peer as a beneficiary
+// The serial and amount of the cheque will depend on the last cheque and current balance for this peer
+// The cheque will be signed and point to the issuer's contract
+func (s *Swap) createCheque(peer enode.ID) (*Cheque, error) {
+	var cheque *Cheque
+	var err error
+
+	swapPeer := s.getPeer(peer)
+	beneficiary := swapPeer.beneficiary
+
+	peerBalance := s.balances[peer]
+	// the balance should be negative here, we take the absolute value:
+	honey := uint64(-peerBalance)
+
+	// convert honey to ETH
+	var amount uint64
+	amount, err = s.oracle.GetPrice(honey)
+	if err != nil {
+		log.Error("error getting price from oracle", "err", err)
+		return nil, err
+	}
+
+	// we need to ignore the error check when loading from the StateStore,
+	// as an error might indicate that there is no existing cheque, which
+	// could mean it's the first interaction, which is absolutely valid
+	_ = s.loadCheque(peer)
```

Comment:
What should happen in case of an error then?

Reply:
Unfortunately ` 72 func (s *DBStore) Get(key string, i interface{}) (err error) {` the only possible error returned is ` return ErrNotFound`, in which case we actually need to ignore it.

I will explicitly change to check that if it is not `ErrNotFound` then it would return the error, but that is currently never happening due to the underlying implementation